README.md
A flutter widget to show the user if there is an update available of your app from an api endpoint.

## Features

* A widget to show if you have an update of your app
* Note that this is an `widget` not a page (not a route) so use this widget as an overlay, (see the
  example at bottom)

## Getting started

No extra dependency is required.  
Just install the package and you are ready to go.

## Usage

Use The provided `ForceUpdatePrompt()` widget to show if your app has an update.

Usage:

##### Example:

```plaintext
ForceUpdatePrompt(
                    androidVersionName: "2.33.4533,
                    androidVersionCode: "22,
                    iosVersionName: "4.34.5434,
                    iosVersionCode: "43,
                    androidPriorityIsHigh: true,
                    iosPriorityIsHigh: false,
                    backgroundColor: Colors.purple,
                    logo: Image.asset("asset/new/icon_app.png"),
                    transparentBackground: true,
                    message: "New version available",
                    updateButton: Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 40.0),
                      child: MyButton(
                        text: "Update Now",
                      ),
                    ),
                    skipButton: Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 40.0),
                      child: MyButton(
                        text: "Skip",
                      ),
                    ),
                  );
```

## Implementation of `validate()`

**It has an helper** `**ForceUpdateUtils.getRemoteUpdateStatus()**` **method if you want to fetch
latest version info from an API endpoint**

##### Example

```plaintext
ForceUpdateUtils.getRemoteUpdateStatus("https://62b43f19530b26da4cb9fcc0.mockapi.io/check-update")
```

If you want to use this build in helper you response must look like this

```json
[
  {
    "message": "New Update is available!",
    "android_version_name": "1.1.20220624",
    "android_version_code": "17",
    "android_priority_is_high": false,
    "ios_version_name": "1.1.20220624",
    "ios_version_code": "17",
    "ios_priority_is_high": true
  }
]
```

As the helper using a internal data model class for this specific JSON formate

**A Full Example**  
Notice that the widget is used as an overlay

### If you want to make the update decision on client side.

```dart
Stack(
          children: [
            isLoading
                ? Splash()
                : isRegistered
                    ? Home()
                    : isAuth
                        ? ProfileNameImageDOB()
                        : IntroPage(),

            FutureBuilder(
              future: ForceUpdateUtils.getRemoteUpdateStatus("https://62b43f19530b26da4cb9fcc0.mockapi.io/check-update"),
              builder: (context, AsyncSnapshot<ForceUpdateRemoteData?> snapshot){
                if(snapshot.hasData && snapshot.data != null){
                  return ForceUpdatePrompt(
                    androidVersionName: snapshot.data!.androidVersionName,
                    androidVersionCode: snapshot.data!.androidVersionCode,
                    iosVersionName: snapshot.data!.iosVersionName,
                    iosVersionCode: snapshot.data!.iosVersionCode,
                    androidPriorityIsHigh: snapshot.data!.androidPriorityIsHigh,
                    iosPriorityIsHigh: snapshot.data!.iosPriorityIsHigh,
                    backgroundColor: Colors.purple,
                    logo: Image.asset("asset/new/icon_app.png"),
                    transparentBackground: true,
                    message: snapshot.data!.message ?? "New version available",
                    updateButton: Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 40.0),
                      child: MyButton(
                        text: "Update Now",
                      ),
                    ),
                    skipButton: Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 40.0),
                      child: MyButton(
                        text: "Skip",
                      ),
                    ),
                  );
                }
                return const Center(child: CircularProgressIndicator());
              } ,
            ),
          ],
        );
```

### If you want to make the update decision on server side.

```dart
FutureBuilder(
  future: ForceUpdateUtils.postLocalInfoForRemoteDecision(
  "https://youryrlhere.com/check-for-update"),
  builder: (context, sn) {
    if (sn.hasData && sn.data != null) {
      ForceUpdateRemoteDecision data = sn.data as ForceUpdateRemoteDecision;
      return UpdateWidget(
        transparentBackground: true,
        backgroundColor: Colors.red,
        logo: Image.network("https://picsum.photos/200"),
        message: data.message ?? "New version available",
        textColor: Colors.purple,
        updateButton: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 40.0),
            child: MyButton(
            text: "Update Now",
          ),
        ),
        skipButton: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 40.0),
          child: MyButton(
          text: "Skip",
        ),
      ), 
      isPriorityHigh: data.isMajor ?? false,);
    }
    return Text("Loading");
    },
)
```

## Additional information
If you want to contribute to the project please go to our github
repo [GitHub](https://github.com/niamulhasan/flutter_force_update_package)

CHANGELOG.md
## 0.0.7

* IOS App Store ID Param Added, and Should show review added


lib/widgets/update_widget.dart
import 'package:flutter/material.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:url_launcher/url_launcher.dart';

import 'package:launch_review/launch_review.dart';

class UpdateWidget extends StatefulWidget {
  const UpdateWidget({
    Key? key,
    required this.transparentBackground,
    required this.backgroundColor,
    required this.logo,
    required this.message,
    required this.textColor,
    required this.updateButton,
    required this.isPriorityHigh,
    required this.skipButton,
    required this.iosAppId,
    required this.shouldWriteReview,
  }) : super(key: key);

  final bool transparentBackground;
  final Color backgroundColor;
  final Widget logo;
  final String message;
  final Color textColor;
  final Widget updateButton;
  final bool isPriorityHigh;
  final Widget skipButton;
  final String iosAppId;
  final bool shouldWriteReview;

  @override
  State<UpdateWidget> createState() => _UpdateWidgetState();
}

class _UpdateWidgetState extends State<UpdateWidget> {
  bool _shouldShow = true;

  @override
  Widget build(BuildContext context) {
    return _shouldShow
        ? Material(
            color: widget.transparentBackground
                ? widget.backgroundColor.withOpacity(0.8)
                : widget.backgroundColor,
            child: Container(
              width: double.infinity,
              height: double.infinity,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Expanded(
                    flex: 6,
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.end,
                      children: [
                        SizedBox(width: 140, child: widget.logo),
                      ],
                    ),
                  ),
                  Expanded(
                    flex: 5,
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.end,
                      children: [
                        Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 8.0),
                          child: Text(
                            widget.message,
                            style: TextStyle(
                                color: widget.textColor,
                                fontSize: 16,
                                fontWeight: FontWeight.w500),
                            textAlign: TextAlign.center,
                          ),
                        ),
                        const SizedBox(height: 14),
                        InkWell(
                          onTap: () async {
                            PackageInfo packageInfo =
                                await PackageInfo.fromPlatform();
                            String packageName = packageInfo.packageName;
                            // final Uri _uri = Uri.parse(
                            //     "https://play.google.com/store/apps/details?id=$packageName");
                            // if (!await launchUrl(_uri)) {
                            //   print("Could not launch url");
                            // }

                            LaunchReview.launch(
                              androidAppId: "$packageName",
                              iOSAppId: widget.iosAppId,
                              writeReview: widget.shouldWriteReview,
                            );
                          },
                          child: widget.updateButton,
                        ),
                        const SizedBox(height: 14),
                        !widget.isPriorityHigh
                            ? InkWell(
                                onTap: () {
                                  setState(() {
                                    _shouldShow = false;
                                  });
                                },
                                child: widget.skipButton,
                              )
                            : Container(),
                        const SizedBox(height: 40),
                      ],
                    ),
                  )
                ],
              ),
            ),
          )
        : Container();
  }
}


lib/utils/force_update_utils.dart
import 'dart:convert';
import 'dart:io';

import 'package:package_info_plus/package_info_plus.dart';

import '../models/force_update_local_data.dart';
import '../models/force_update_remote_data.dart';
import 'package:http/http.dart' as http;

import '../models/force_update_remote_decision.dart';

class ForceUpdateUtils {
  static Future<ForceUpdateRemoteData?> getRemoteUpdateStatus(String url) async {
    http.Response response = await http.get(Uri.parse(url));
    if (response.statusCode == 200) {
      return ForceUpdateRemoteData.fromJson(jsonDecode(response.body)[0]);
    }
    return null;
  }

  static Future<ForceUpdateRemoteDecision> postLocalInfoForRemoteDecision(
      String url) async {
    PackageInfo packageInfo = await PackageInfo.fromPlatform();
    ForceUpdateLocalData localData = ForceUpdateLocalData(
        platform: Platform.isAndroid ? "android" : "ios",
        appName: "romoni-user",
        versionCode: packageInfo.buildNumber,
    );
    http.Response response = await http.post(Uri.parse(url),
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(localData.toJson()));
    if (response.statusCode == 200) {
      return ForceUpdateRemoteDecision.fromJson(jsonDecode(response.body));
    }
    return ForceUpdateRemoteDecision();
  }
}

lib/widgets/force_update_prompt.dart
part of flutter_force_update;

class ForceUpdatePrompt extends StatelessWidget {
  final String? androidVersionName;
  final String? androidVersionCode;
  final String? iosVersionName;
  final String? iosVersionCode;
  final bool? androidPriorityIsHigh;
  final bool? iosPriorityIsHigh;
  final bool ignore;
  final Color backgroundColor;
  final Color textColor;
  final Widget logo;
  final bool transparentBackground;
  final String message;
  final Widget updateButton;
  final Widget skipButton;

  const ForceUpdatePrompt({
    Key? key,
    required this.androidVersionName,
    required this.androidVersionCode,
    required this.iosVersionName,
    required this.iosVersionCode,
    required this.androidPriorityIsHigh,
    required this.iosPriorityIsHigh,
    this.ignore = false,
    required this.backgroundColor,
    this.textColor = Colors.white,
    required this.logo,
    this.transparentBackground = false,
    required this.message,
    required this.updateButton,
    required this.skipButton,
  }) : super(key: key);

  bool get isPriorityHigh {
    if (Platform.isAndroid) {
      return androidPriorityIsHigh ?? false;
    } else {
      return iosPriorityIsHigh ?? false;
    }
  }

  Future<bool> _isUpdateAvailable() async {
    PackageInfo packageInfo = await PackageInfo.fromPlatform();
    // String appName = packageInfo.appName;
    // String packageName = packageInfo.packageName;
    int localVersion = int.parse(packageInfo.version.replaceAll(".", ""));
    int localBuildNumber = int.parse(packageInfo.buildNumber);

    // LOGGING START
    if (false) {
      print(
          "Version Name Android: Local:Remote : $localVersion : $androidVersionName");
      print("Version Name iOS: Local:Remote : $localVersion : $iosVersionName");
      print(
          "Version Code Android: Local:Remote : $localBuildNumber : $androidVersionCode");
      print(
          "Version Code iOS: Local:Remote : $localBuildNumber : $iosVersionName");
    }
    //LOGGING END

    if (Platform.isAndroid) {
      if (androidVersionName == null || androidVersionCode == null) {
        return false;
      }
      if (localVersion < int.parse(androidVersionName!.replaceAll(".", "")) ||
          localBuildNumber < int.parse(androidVersionCode!)) {
        return true;
      }
    }

    if (Platform.isIOS) {
      if (iosVersionName == null || iosVersionCode == null) {
        return false;
      }
      if (localVersion < int.parse(iosVersionName!.replaceAll(".", "")) ||
          localBuildNumber < int.parse(iosVersionCode!)) {
        return true;
      }
    }
    return false;
  }

  void printIsUpdateAvailable() async {
    print("Update decision: " + (await _isUpdateAvailable()).toString());
  }

  @override
  Widget build(BuildContext context) {
    printIsUpdateAvailable();
    return !ignore
        ? FutureBuilder(
            future: _isUpdateAvailable(),
            builder: (context, AsyncSnapshot<bool> snapshot) {
              if (snapshot.hasData && (snapshot.data ?? false)) {
                return UpdateWidget(
                  transparentBackground: transparentBackground,
                  backgroundColor: backgroundColor,
                  logo: logo,
                  message: message,
                  textColor: textColor,
                  updateButton: updateButton,
                  isPriorityHigh: isPriorityHigh,
                  skipButton: skipButton,
                );
              } else {
                return Container();
              }
              return const Center(
                  child: CircularProgressIndicator(color: Colors.red));
            },
          )
        : Container();
  }
}


lib/models/force_update_remote_data.dart
class ForceUpdateRemoteData {
  String? androidVersionName;
  String? androidVersionCode;
  String? iosVersionName;
  String? iosVersionCode;
  bool? androidPriorityIsHigh;
  bool? iosPriorityIsHigh;
  String? message;

  ForceUpdateRemoteData({
    this.androidVersionName,
    this.androidVersionCode,
    this.iosVersionName,
    this.iosVersionCode,
    this.androidPriorityIsHigh,
    this.iosPriorityIsHigh,
    this.message,
  });

  ForceUpdateRemoteData.fromJson(Map<String, dynamic> json) {
    androidVersionName = json['android_version_name'];
    androidVersionCode = json['android_version_code'];
    iosVersionName = json['ios_version_name'];
    iosVersionCode = json['ios_version_code'];
    androidPriorityIsHigh = json['android_priority_is_high'];
    iosPriorityIsHigh = json['ios_priority_is_high'];
    message = json['message'];
  }
}


example/main.dart
// import 'package:flutter/material.dart';
// import 'package:flutter_force_update/flutter_force_update.dart';
// import 'package:flutter_force_update/models/force_update_local_data.dart';
// import 'package:flutter_force_update/models/force_update_remote_data.dart';
// import 'package:flutter_force_update/models/force_update_remote_decision.dart';
// import 'package:flutter_force_update/utils/force_update_utils.dart';
// import 'package:flutter_force_update/widgets/update_widget.dart';
//
// void main() => runApp(const MyApp());
//
// class MyApp extends StatelessWidget {
//   const MyApp({super.key});
//
//   @override
//   Widget build(BuildContext context) {
//     return MaterialApp(
//       title: 'Material App',
//       home: Scaffold(
//         appBar: AppBar(
//           title: const Text('Material App Bar'),
//         ),
//         body: HomePage(),
//       ),
//     );
//   }
// }
//
// class HomePage extends StatelessWidget {
//   const HomePage({
//     Key? key,
//   }) : super(key: key);
//
//   @override
//   Widget build(BuildContext context) {
//     return Stack(
//       children: [
//         Center(
//           child: Container(
//             child: Text("Hello"),
//           ),
//         ),
//         // FutureBuilder(
//         //   future: ForceUpdateUtils.getRemoteUpdateStatus(
//         //       "https://62b43f19530b26da4cb9fcc0.mockapi.io/check-update"),
//         //   builder: (context, AsyncSnapshot<ForceUpdateRemoteData?> snapshot) {
//         //     if (snapshot.hasData && snapshot.data != null) {
//         //       return ForceUpdatePrompt(
//         //         androidVersionName: snapshot.data!.androidVersionName,
//         //         androidVersionCode: snapshot.data!.androidVersionCode,
//         //         iosVersionName: snapshot.data!.iosVersionName,
//         //         iosVersionCode: snapshot.data!.iosVersionCode,
//         //         androidPriorityIsHigh: snapshot.data!.androidPriorityIsHigh,
//         //         iosPriorityIsHigh: snapshot.data!.iosPriorityIsHigh,
//         //         backgroundColor: Colors.purple,
//         //         logo: Icon(
//         //           Icons.handyman,
//         //           size: 140,
//         //         ),
//         //         transparentBackground: true,
//         //         message: snapshot.data!.message ?? "New version available",
//         //         updateButton: Padding(
//         //           padding: const EdgeInsets.symmetric(horizontal: 40.0),
//         //           child: MyButton(
//         //             text: "Update Now",
//         //           ),
//         //         ),
//         //         skipButton: Padding(
//         //           padding: const EdgeInsets.symmetric(horizontal: 40.0),
//         //           child: MyButton(
//         //             text: "Skip",
//         //           ),
//         //         ),
//         //       );
//         //     }
//         //     return const Center(child: CircularProgressIndicator());
//         //   },
//         // ),
//
//
//         //For server side decision
//
//
//         FutureBuilder(
//           future: ForceUpdateUtils.postLocalInfoForRemoteDecision(
//               "https://developers.romoni.com.bd/api/v3.0/check-version"),
//           builder: (context, sn) {
//             if (sn.hasData && sn.data != null) {
//               ForceUpdateRemoteDecision data =
//               sn.data as ForceUpdateRemoteDecision;
//               return UpdateWidget(
//                 transparentBackground: true,
//                 backgroundColor: Colors.red,
//                 logo: Image.network("https://picsum.photos/200"),
//                 message: data.message ?? "New version available",
//                 textColor: Colors.purple,
//                 updateButton: Padding(
//                   padding: const EdgeInsets.symmetric(horizontal: 40.0),
//                   child: MyButton(
//                     text: "Update Now",
//                   ),
//                 ),
//                 skipButton: Padding(
//                   padding: const EdgeInsets.symmetric(horizontal: 40.0),
//                   child: MyButton(
//                     text: "Skip",
//                   ),
//                 ), isPriorityHigh: data.isMajor ?? false,);
//             }
//             return Text("Loading");
//           },
//         ),
//       ],
//     );
//   }
// }
//
// class MyButton extends StatelessWidget {
//   final String text;
//
//   const MyButton({Key? key, required this.text}) : super(key: key);
//
//   @override
//   Widget build(BuildContext context) {
//     return Container(color: Colors.red, child: Text(text));
//   }
// }
